pkg/activator/net/throttler.go

```diff
@@ -0,0 +1,489 @@
+	"knative.dev/serving/pkg/reconciler"
+	"knative.dev/serving/pkg/resources"
+)
+
+type podIPTracker struct {
+	dest     string
+	requests int
+}
+
+type breaker interface {
+	Capacity() int
+	Maybe(ctx context.Context, thunk func()) bool
+	UpdateConcurrency(int) error
+}
+
+type revisionThrottler struct {
+	revID                types.NamespacedName
+	containerConcurrency int64
+
+	// Holds the current number of backends. This is used for when we get an activatorCount update and
+	// therefore need to recalculate capacity
+	backendCount int
+
+	// This is a breaker for the revision as a whole. try calls first pass through
+	// this breaker and are either called with clusterIPDest or go through selecting
+	// a podIPTracker and are then called.
+	breaker breaker
+
+	// If we have a healthy clusterIPDest this is set to nil.
+	podIPTrackers []*podIPTracker
+
+	// If we dont have a healthy clusterIPDest this is set to the default, otherwise
+	// it is the l4dest for this revision's private clusterIP
+	clusterIPDest string
+
+	// mux guards "throttle state" which is the state we use during the request path. This
+	// is trackers, clusterIPDest
+	mux sync.RWMutex
+
+	// used to atomically calculate and set capacity
+	capacityMux sync.Mutex
+
+	logger *zap.SugaredLogger
+}
+
+func newRevisionThrottler(revID types.NamespacedName,
+	containerConcurrency int64,
+	breakerParams queue.BreakerParams,
+	logger *zap.SugaredLogger) *revisionThrottler {
+	var revBreaker breaker
+	if containerConcurrency == 0 {
+		revBreaker = NewInfiniteBreaker()
+	} else {
+		revBreaker = queue.NewBreaker(breakerParams)
+	}
+	return &revisionThrottler{
+		revID:                revID,
+		containerConcurrency: containerConcurrency,
+		breaker:              revBreaker,
+		logger:               logger,
+	}
+}
+
+// Returns clusterIP if it is the current valid dest. If neither clusterIP or podIPs are valid
+// dests returns error.
+func (rt *revisionThrottler) checkClusterIPDest() (string, error) {
+	rt.mux.RLock()
+	defer rt.mux.RUnlock()
+
+	if rt.clusterIPDest == "" && len(rt.podIPTrackers) == 0 {
+		return "", errors.New("made it through breaker but we have no clusterIP or podIPs. This should" +
+			" never happen" + rt.revID.String())
+	}
+
+	return rt.clusterIPDest, nil
+}
+
+// Returns a dest after incrementing it's request cnt and a completion callback
+// to be called after request completion
+func (rt *revisionThrottler) acquireDest() (string, func()) {
+	rt.mux.Lock()
+	defer rt.mux.Unlock()
+
+	if rt.clusterIPDest != "" {
+		return rt.clusterIPDest, func() {}
+	}
+
+	var leastConn *podIPTracker
+	// Find the dest with fewest active connections.
+	for _, tracker := range rt.podIPTrackers {
+		if leastConn == nil || leastConn.requests > tracker.requests {
+			leastConn = tracker
+		}
+	}
+
+	if leastConn != nil {
+		leastConn.requests++
+		return leastConn.dest, func() {
+			rt.mux.Lock()
+			defer rt.mux.Unlock()
+			leastConn.requests--
+		}
+	}
+	return "", func() {}
```

Comment:
I'd return `nil` for the func, and below rewrote as
```
If r =="" { ret=errors(); return}
defer cb()
ret = f()
```
Also follows best practice (https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow)